flink-ml-iteration/src/main/java/org/apache/flink/iteration/datacache/nonkeyed/DataCacheWriter.java

```diff
@@ -19,127 +19,158 @@
+        this.segmentPool = segmentPool;
+        this.serializer = serializer;
+        this.finishedSegments = new ArrayList<>();
+        this.finishedSegments.addAll(finishedSegments);
+        this.currentWriter = createSegmentWriter();
     }
 
-    public List<Segment> finish() throws IOException {
-        finishCurrentSegment(false);
-        return finishSegments;
+    public void addRecord(T record) throws IOException {
+        assert currentWriter != null;
+        if (!currentWriter.addRecord(record)) {
+            currentWriter.finish().ifPresent(finishedSegments::add);
+            currentWriter = new FileSegmentWriter<>(serializer, pathGenerator.get());
+            Preconditions.checkState(currentWriter.addRecord(record));
+        }
     }
 
-    public FileSystem getFileSystem() {
-        return fileSystem;
-    }
+    /** Finishes current segment if records has ever been added to this segment. */
+    public void finishCurrentSegmentIfAny() throws IOException {
+        if (currentWriter == null || currentWriter.getCount() == 0) {
+            return;
+        }
 
-    public List<Segment> getFinishSegments() {
-        return finishSegments;
+        currentWriter.finish().ifPresent(finishedSegments::add);
+        currentWriter = createSegmentWriter();
     }
 
-    private void finishCurrentSegment(boolean newSegment) throws IOException {
-        if (currentSegment != null) {
-            currentSegment.finish().ifPresent(finishSegments::add);
-            currentSegment = null;
+    /** Finishes adding records and closes resources occupied for adding records. */
+    public List<Segment> finish() throws IOException {
+        if (currentWriter == null) {
+            return finishedSegments;
         }
 
-        if (newSegment) {
-            currentSegment = new SegmentWriter(pathGenerator.get());
-        }
+        currentWriter.finish().ifPresent(finishedSegments::add);
+        currentWriter = null;
+        return finishedSegments;
     }
 
-    private class SegmentWriter {
-
-        private final Path path;
-
-        private final FSDataOutputStream outputStream;
-
-        private final DataOutputView outputView;
-
-        private int currentSegmentCount;
+    public List<Segment> getFinishedSegments() {
+        return finishedSegments;
+    }
 
-        public SegmentWriter(Path path) throws IOException {
-            this.path = path;
-            this.outputStream = fileSystem.create(path, FileSystem.WriteMode.NO_OVERWRITE);
-            this.outputView = new DataOutputViewStreamWrapper(outputStream);
+    /** Cleans up all previously added records. */
+    public void cleanup() throws IOException {
+        finishCurrentSegmentIfAny();
+        for (Segment segment : finishedSegments) {
+            if (segment.isOnDisk()) {
+                fileSystem.delete(segment.getPath(), false);
+            }
+            if (segment.isCached()) {
+                assert segmentPool != null;
```

Comment:
After thinking about this more, I find it more consistent with other code to just remove the assert here. Note that the code will throw NullPointerException if segmentPool == null, which is good enough for us to investigate this bug.